### ISSUE TITLE
Emergent behaviour of relentless increase

### ISSUE THREAD

**Issue by keithreid-sfw** [tobiasbrodd/neural-network-julia#1]
Dear Tobias

Thankyou for your code. I have a question which may or may not be an issue. I see your code as a base for a project.  I have altered the code but you’ll recognise aspects. I am doing economics. If I get published I’ll cite your GitHub, or cite you in some other manner if you send your preferred citation format. 

Paradigm is as follows.  A set of neural networks based on your code interact and I plot that.

Issue is as follows. I have noticed that when I plot the output of the neural network that output values seem to rise, all else being equal. They can go down, but as they become stable in their differentials they rise. 

Is this a known feature?  If not it might be emergent behaviour between networks. I have been weighing up whether to try to remove it. I think it is a property of the code, which is fine, I just don’t want to be missing something. 

Looking for a cause I have been wondering if the fact that epsilon is a small positive value and is always .+= or += added leads to a tendency for the output to increase. Is this dumb? I was thinking about ways to mitigate it.

What do you think? Here is my implementation, you’ll recognise it. You’ll see the output is three values, called eastness, upness, and choice. I used ‘east’ and ‘up’ as terms to avoid clashes with X and Y!  🙂 

Most of the other new stuff is just record keeping so I can plot at the end of my big code. You’ll also see the new function X_y_result_to_utility_alice(X_alice, y, choice_alice) which doesn’t contribute the emergent behaviour, it’s just solving the problem the nets are faced with. 

Finally, and before ending my question, in order to avoid this rise, I want to make  either foo=-1 or foo=+1, control that by alternating between one and other for regular blocks of iteration, or making the choice random, and do something like:

`network_alice.W[end]` += network_alice.epsilon * foo * transpose(network_alice.a[end-1]) * delta_alice
    network_alice.b[end] .+= network_alice.epsilon * foo * `mean(delta_alice)`

One issue is wrecking the gradient descent, which is why I am considering having blocks of ten iterations of foo=-1 then ten foo=+1, I am going to try this anyway but I thought you’d have a perspective. 

Thanks, I’ll sign off here. 

So, to finish, here’s the existing, developed implementation:

`function backward_alice!(network_alice::Network_alice, X_alice, 
        y, record_of_easts_alice, record_of_ups_alice, record_of_choices_alice, record_of_utilities_alice)
    choice_alice = reshape(network_alice.result, 3)
    eastness_alice = choice_alice[1]
    upness_alice = choice_alice[2]
    choice_alice = choice_alice[3]
    
    append!(record_of_easts_alice, eastness_alice)
    append!(record_of_ups_alice, upness_alice)
    
    append!(record_of_choices_alice, choice_alice)
    utility_alice = X_y_result_to_utility_alice(X_alice, y, choice_alice)
    error_alice = y - utility_alice
    append!(record_of_utilities_alice, utility_alice)
    delta_alice = error_alice .* sigmoid_derivative(network_alice.result)
    network_alice.W[end] += network_alice.epsilon * transpose(network_alice.a[end-1]) * delta_alice
    network_alice.b[end] .+= network_alice.epsilon * mean(delta_alice)
    W_length = length(network_alice.W)
    for i = 1:W_length-1
        j = W_length - i   
        error_alice = delta_alice * transpose(network_alice.W[j+1])
        delta_alice = error_alice .* sigmoid_derivative(network_alice.a[j+1])
        network_alice.W[j] += network_alice.epsilon * transpose(network_alice.a[j]) * delta_alice
        network_alice.b[j] .+= network_alice.epsilon * mean(delta_alice)
    end
    return eastness_alice, upness_alice, choice_alice, network_alice, record_of_easts_alice, record_of_ups_alice, record_of_choices_alice, record_of_utilities_alice
end`

Edit:  I just flipped the values to negatives in the setup and it did flip the behaviour. I’ll go for the blocks of alternating polarity. 

Still interested in your thoughts.

**Comment by keithreid-sfw** [comment]
And here is a python mockup I did for the control block; I’ll let you know how it goes; the idea I that for each back_propagation I multiply epsilon for all nets by this polarity. 

Python script:

def block_loop(learning_opportunities, max_block):
    polarity = 1
    block = 0
    for i in range(learning_opportunities):
        block = block +1
        print(polarity)
        if block%max_block == 0:
            polarity = polarity *(-1)
           
block_loop(1000,1)
